std/range/package.d

```diff
@@ -2465,6 +2465,114 @@ Take!(Repeat!T) repeat(T)(T value, size_t n)
 }
 
 /**
+Given callable ($(XREF traits, isCallable)) $(D fun), create as a range
+whose front is defined by successive calls to $(D fun()).
+This is especially useful to call function with global side effects (random
+functions), or to create ranges expressed as a single delegate, rather than
+an entire $(D front)/$(D popFront)/$(D empty) structure.
+$(D fun) maybe be passed either a template alias parameter (existing
+function, delegate, struct type defining static $(D opCall)... ) or
+a run-time value argument (delegate, function object... ).
+The result range models an InputRange ($(XREF range, isInputRange).
+The resulting range will call $(D fun()) on every call to $(D front),
+and only when $(D front) is called, regardless of how the range is
+iterated. 
+It is advised to compose with either $(XREF algorithm,cache) or 
+$(XREF array,array), or to use in a (by-value) foreach loop.
+
+Params:
+    Fun = is the $(D isCallable) that gets called on every call to front.
+
+Returns: an $(D inputRange) that returns a new value generated by $(D Fun) on
+    any call to $(D front).
+*/
+auto generate(Fun)(Fun fun)
+    if (isCallable!fun)
+{
+    return Generator!(true, Fun)(fun);
+}
+
+///
+auto generate(alias fun)()
+    if (isCallable!fun)
+{
+    return Generator!(false, fun)();
+}
+
+//private struct Generator(bool onPopFront, bool runtime, Fun...)
+private struct Generator(bool runtime, Fun...)
+{
+    static assert(Fun.length == 1);
+    static assert(isInputRange!Generator);
+
+private:
+    static if (runtime)
+        Fun[0] fun;
+    else
+        alias fun = Fun[0];
```

Comment:
Could you use `static if (is(Fun[0]))` instead of the extra `runtime bool`?
In any case this way of function passing is complex boilerplate, we should come up with a nicer idiom at some point.